src/dbnode/storage/index/block_prop_test.go

```diff
@@ -197,3 +206,207 @@ func newPropTestBlock(t *testing.T, blockStart time.Time, nsMeta namespace.Metad
+					Value: []byte(val),
+				})
+			}
+		}
+
+		return propTestSegment{
+			metadata:   doc.Metadata{Fields: docFields},
+			exCount:    int64(len(segMap)),
+			segmentMap: segMap,
+		}
+	})
+}
+
+func verifyResults(
+	t *testing.T,
+	results AggregateResults,
+	exMap segmentMap,
+) {
+	resultMap := make(segmentMap, results.Map().Len())
+	for _, field := range results.Map().Iter() { //nolint:gocritic
+		name := field.Key().String()
+		_, found := resultMap[name]
+		require.False(t, found, "duplicate values in results map")
+
+		values := make(testValuesSet, field.value.Map().Len())
+		for _, value := range field.value.Map().Iter() {
+			val := value.Key().String()
+			_, found := values[val]
+			require.False(t, found, "duplicate values in results map")
+
+			values[val] = struct{}{}
+		}
+
+		resultMap[name] = values
+	}
+
+	require.Equal(t, resultMap, exMap)
+}
+
+func TestAggregateDocLimits(t *testing.T) {
+	var (
+		parameters = gopter.DefaultTestParameters()
+		seed       = time.Now().UnixNano()
+		reporter   = gopter.NewFormatedReporter(true, 160, os.Stdout)
+	)
+
+	parameters.MinSuccessfulTests = 1000
+	parameters.MinSize = 5
+	parameters.MaxSize = 10
+	parameters.Rng = rand.New(rand.NewSource(seed)) //nolint:gosec
+	properties := gopter.NewProperties(parameters)
+
+	properties.Property("segments dedupe and have correct docs counts", prop.ForAll(
+		func(testSegment propTestSegment) (bool, error) {
+			seg, err := mem.NewSegment(mem.NewOptions())
+			if err != nil {
+				return false, err
+			}
+
+			_, err = seg.Insert(testSegment.metadata)
+			if err != nil {
+				return false, err
+			}
+
+			err = seg.Seal()
+			if err != nil {
+				return false, err
+			}
+
+			scope := tally.NewTestScope("", nil)
+			iOpts := instrument.NewOptions().SetMetricsScope(scope)
+			limitOpts := limits.NewOptions().
+				SetInstrumentOptions(iOpts).
+				SetDocsLimitOpts(limits.LookbackLimitOptions{Lookback: time.Minute}).
+				SetBytesReadLimitOpts(limits.LookbackLimitOptions{Lookback: time.Minute})
+			queryLimits, err := limits.NewQueryLimits((limitOpts))
+			require.NoError(t, err)
+			testOpts = testOpts.SetInstrumentOptions(iOpts).SetQueryLimits(queryLimits)
+
+			testMD := newTestNSMetadata(t)
+			start := time.Now().Truncate(time.Hour)
+			blk, err := NewBlock(start, testMD, BlockOptions{},
+				namespace.NewRuntimeOptionsManager("foo"), testOpts)
+			if err != nil {
+				return false, err
+			}
+
+			b, ok := blk.(*block)
+			if !ok {
+				return false, errors.New("bad block type")
+			}
+
+			b.mutableSegments.foregroundSegments = []*readableSeg{
+				newReadableSeg(seg, testOpts),
+			}
+
+			results := NewAggregateResults(ident.StringID("ns"), AggregateResultsOptions{
+				Type: AggregateTagNamesAndValues,
+			}, testOpts)
+
+			ctx := context.NewContext()
+			defer ctx.BlockingClose()
+
+			// create initial span from a mock tracer and get ctx
+			ctx.SetGoContext(opentracing.ContextWithSpan(stdlibctx.Background(), nil))
```

Comment:
ah interesting, why do we need to create a span for the test?

Reply:
Ah damn good catch, thought I'd removed this, the test I used as a base did some verification on tracepoints